### ISSUE TITLE
karpenter nodes launch on skewed subnets

### ISSUE THREAD

**Issue by zensai3805** [aws/karpenter#3364]
### Version

__Karpenter Version:__ v0.23.0

__Kubernetes Version:__ v1.24.0

### Expected Behavior

Multiple subnets are associated with one AZ.
Nodes are distributed across multiple subnets when k8s pods are scaled out on a large scale. Each subnet has an even number of available IPs.

According to the document, Karpenter selects a subnet with many available IPs and laundh nodes.

### Actual Behavior

When scaling out a node, Karpenter looks at the number of available IPs on a subnet only once and chooses which subnet to start the node on.

As a result, nodes are started on a skewed subnet. If you scale out a lot at once, you'll run out of IPs on certain subnets, leaving certain subnets underutilized.

### Steps to Reproduce the Problem

1) Create multiple subnets in an AZ and configure AWSNodeTemplate so that karpenter can launch nodes on each subnet.
2) Scale out your pods. Even if you scale out to the point of exhausting a subnet's IPs, the pods are not distributed across subnets.

### Resource Specs and Logs

Log when IPs on subnet are exhausted
```
Failed to create pod sandbox: rpc error: code = Unknown desc = failed to setup network for sandbox "xxxxxxx": plugin type="aws-cni" 
│  name="aws-cni" failed (add): add cmd: failed to assign an IP address to container
```

AWSNodeTemplate
```
apiVersion: karpenter.k8s.aws/v1alpha1
kind: AWSNodeTemplate
metadata:
  name: karpenter-node-template
spec:
  subnetSelector:
    aws-ids: "subnet-xxxxx,subnet-yyyy"
  securityGroupSelector:
    aws-ids: "sg-xxxx"
  tags:
    ManagedByKarpenter: "true"
```

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

**Comment by bwagner5** [comment]
I think we can do a better job at tracking IP addresses, but we'll never be perfect since we won't have perfect information on how the CNI is configured. For example, we'll use less IPs if using an overlay network rather than a subnet IP per pod. 

The approach I'd take for implementation is to track IPs per subnet in-memory and do an opportunistic deduction of IPs when we're constructing the launch request to EC2. Since the EC2 request can include multiple subnets and instance types, we'll need to deduct from subnets that may not actually consume the IPs. But we can re-add the IPs once the ec2 request completes and we get a final result. I think we'll also want to delete the in-memory IP tracking after the subnet provider's TTL, since EC2 will return to us updated IP counts.

I may take a crack at this implementation and see if it works well since I've heard this request a few times.

**Comment by FernandoMiguel** [comment]
@bwagner5 why not query the subnets attached to the `AWSNodeTemplate` and do a describe from aws api, once all the recently launched nodes are in a `ready` state?
that should return you the exact IPs available, even if /28 are attached to ENIs.

**Comment by bwagner5** [comment]
> @bwagner5 why not query the subnets attached to the `AWSNodeTemplate` and do a describe from aws api, once all the recently launched nodes are in a `ready` state? that should return you the exact IPs available, even if /28 are attached to ENIs.

We'll do that when the subnet cache expires to get an exact update which is 1 minutes. The tracking mechanism I'm proposing is just predicting IPs while we're launching nodes based on the number of pods that could be packed on the nodes. This will prevent launching all the nodes into the subnet with the most IPs even when you've already launched so many nodes that it is no longer the subnet with the most IPs.

**Comment by bwagner5** [comment]
Just merged the inflight IP tracking for subnets, so should be a better experience. Will be released in the next release v0.26.0. Reopen this issue if you're still seeing this problem after v0.26.0
